staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -201,7 +201,9 @@ func (tw *baseTimeoutWriter) WriteHeader(code int) {
 	tw.mu.Lock()
 	defer tw.mu.Unlock()
 
-	if tw.timedOut || tw.wroteHeader || tw.hijacked {
+	// http.StatusInternalServerError(500) passes through here, regardless of
+	// tw.wroteHeader or tw.hijacked as 500 errors must be propogated for logging (#75983)
+	if (tw.timedOut || tw.wroteHeader || tw.hijacked) && code != http.StatusInternalServerError {
```

Comment:
Do you know why the 500 is passed to this writer, rather than already being written by inner layer of handlers?

Also, how do we know if it's safe to overwrite the existing error code in all cases where 500 is passed in?

Reply:
> Do you know why the 500 is passed to this writer, rather than already being written by inner layer of handlers?

The issue is that there are '500' failure modes where a Write is called, triggering a WriteHeader from our http.ResponseWriter wrappers but the Write fails (eg: erroring [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L103), with the 500 coming from [here](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/json/json.go#L307)).  The inner layer of handlers call a WriteHeader(500) which bubbles up through all of the http.ResponseWriter wrappers but the timeouthandler stops this 500 from bubbling up to httplog from the if condition in this PR.  The original logic to disallow writing the header a second time makes sense as a header cannot be written twice but this 500 case is unique as we want it logged but by short circuiting in the timeout handler, the httplog wrapper never receives a WriteHeader(500) which we want in our logs.

> Also, how do we know if it's safe to overwrite the existing error code in all cases where 500 is passed in?
This [ErrorToAPI code path](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/status.go#L60-L82) was the path for the 500 that this PR used for testing.  I don't think it's possible there is a case where we would overwrite a valid status with a 500 with this change, can you think of a code path where this occurs?  Once a 500 occurs, I don't think much else can be done but log the 500 which is what this PR allows.

The idea of the change in this PR is a 500 "breakglass" for errors that occur on Write so that they are logged via the httplog http.ResponseWriter wrapper. Outside of 500s appearing in httplog, I don't believe there are side effects.